File: Sort/SortPractice.py
# Practice 13. Sorting 
from math import floor
MERGE_SORT = 'M'
QUICK_SORT = 'Q'

def Merge(L1, L2):
  tmp = []
  while L1 and L2:
    if L1[0] > L2[0]:
      tmp.append(L1.pop(0))
    else:
      tmp.append(L2.pop(0))
  l = L1 if L1 else L2
  tmp = tmp + l
  return tmp

def MergeSort(L):
  if len(L) <= 1:
    return L
  mid = len(L) // 2
  L1 = MergeSort(L[:mid])
  L2 = MergeSort(L[mid:])
  tmp = Merge(L1, L2)
  return tmp

def Partition(L, low, high):
  pivot = L[high]
  l = low
  r = high
  while l < r:
    while l < high and L[l] > pivot:
      l += 1
    while r > low and L[r] <= pivot:
      r -= 1
    if l < r:
      L[l], L[r] = L[r], L[l]
  L[l], L[high] = L[high], L[l]
  return l

def QuickSort(L, l, r):
  if l < r:
    p = Partition(L, l, r)
    QuickSort(L, l, p-1)
    QuickSort(L, p+1, r)

if __name__ == "__main__":
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    i = 0
    while i < len(lines):
      line = lines[i].split()
      if len(line) != 2:
        raise Exception("Error: invalid input")
      op = line[0]
      size = int(line[1])
      i += 1
      arr = [int(j) for j in lines[i].split()]

      if op == MERGE_SORT:
        if len(line) != 2:
          raise Exception("MERGE_SORT: invalid input")
        sortedList = MergeSort(arr)
        sortedList = [str(j) for j in sortedList]
        outFile.write(" ".join(sortedList))
        outFile.write("\n")
        
      elif op == QUICK_SORT:
        if len(line) != 2:
          raise Exception("QUICK_SORT: invalid input")
        QuickSort(arr, 0, len(arr)-1)
        sortedList = [str(j) for j in arr]
        outFile.write(" ".join(sortedList))
        outFile.write("\n")

      else:
        raise Exception("Undefined operator")
      
      i += 1

File: Sort/SortPractice.cpp
// Practice 13. Sorting 
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>
using namespace std;
const char MERGE_SORT = 'M';
const char QUICK_SORT = 'Q';

void Merge(vector<int>& L, vector<int>& T, int lPos, int rPos, int rEnd) {
  int lEnd = rPos - 1;
  int tPos = lPos;
  int size = rEnd - lPos + 1;
  while (lPos <= lEnd && rPos <= rEnd) {
    if (L[lPos] > L[rPos]) 
      T[tPos++] = (L[lPos++]);
    else
      T[tPos++] = (L[rPos++]);
  }

  while (lPos <= lEnd)
    T[tPos++] = L[lPos++];
  while (rPos <= rEnd)
    T[tPos++] = L[rPos++];
  for (int i=0; i<size; ++i, --rEnd) {
    L[rEnd] = T[rEnd];
  }
}

void MergeSort(vector<int>& L, vector<int>& T, int left, int right) {
  if (left < right) {
    int mid = (left + right) / 2;
    MergeSort(L, T, left, mid);
    MergeSort(L, T, mid+1, right);
    Merge(L, T, left, mid+1, right);
  }
}

int Partition(vector<int>& L, int low, int high) {
  int pivot = L[high];
  int left = low, right = high;
  
  while (left < right) {
    while (L[left] < pivot)
      left++;
    while (L[right] >= pivot)
      right--;
    if (left < right)
      swap(L[left], L[right]);
  }

  L[high] = L[left];
  L[left] = pivot;
  return left;
}

void QuickSort(vector<int>& L, int low, int high) {
  if (low < high) {
    int p = Partition(L, low, high);
    QuickSort(L, low, p-1);
    QuickSort(L, p+1, high);
  }
}

int main() {
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(2));
    vector<int> arr;
    int size;
    vector<int> tempArray(10, 10);
    // cout <<line<<endl;
    getline(inFile, line);
    istringstream is(line);
    int n;
    while (is >> n)
      arr.push_back(n);

    switch(op)
    {
      case MERGE_SORT:
        // TODO
        MergeSort(arr, tempArray, 0, arr.size()-1);
        for (int i=0; i<arr.size(); i++)
          outFile << arr[i] << " ";
        outFile << endl;
        break;

      case QUICK_SORT:
        // TODO
        QuickSort(arr, 0, arr.size()-1);
        for (int i=0; i<arr.size(); i++)
          outFile << arr[i] << " ";
        outFile << endl;
        break;

      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
